File: dailyQuestion1927.cpp
class Solution {
public:
    bool sumGame(string num) {

        int n = num.size();
        int half = n / 2;

        int sumDiff = 0;
        int questionDiff = 0;

        for (int i = 0; i < half; i++) {

            if (num[i] == '?') {
                questionDiff++;
            } else {
                sumDiff += num[i] - '0';
            }
        }

        for (int i = half; i < n; i++) {

            if (num[i] == '?') {
                questionDiff--;
            } else {
                sumDiff -= num[i] - '0';
            }
        }

        // If number of ? is odd, Alice gets the final move
        int totalQuestions = 0;

        for (char c : num) {
            if (c == '?') {
                totalQuestions++;
            }
        }

        if (totalQuestions % 2 == 1) {
            return true;
        }

        int maxDifference =
            (abs(questionDiff) / 2) * 9;

        return abs(sumDiff) != maxDifference;
    }
};

File: dailyQuestion1927.java
class Solution {
    public boolean sumGame(String num) {

        int n = num.length();
        int half = n / 2;

        int sumDiff = 0;
        int questionDiff = 0;

        for (int i = 0; i < half; i++) {

            if (num.charAt(i) == '?') {
                questionDiff++;
            } else {
                sumDiff += num.charAt(i) - '0';
            }
        }

        for (int i = half; i < n; i++) {

            if (num.charAt(i) == '?') {
                questionDiff--;
            } else {
                sumDiff -= num.charAt(i) - '0';
            }
        }

        /*
         * Bob can win only when:
         *
         * question marks are balanced in a way that
         * allows the remaining sum difference to be zero.
         */
        return sumDiff != -questionDiff * 9 / 2;
    }
}